nogotofail/mitm/connection/handlers/data/ssl.py

```diff
@@ -203,3 +205,52 @@ def on_ssl(self, client_hello):
                 self.log(logging.ERROR,
                         "Client enabled SSLv3 protocol without TLS_FALLBACK_SCSV")
             self.log_attack_event(data="SSLv3")
+
+@handler.passive(handlers)
+class NoForwardSecrecy(_TlsRecordHandler):
+    name = "noforwardsecrecy"
+    description = (
+        "Detects selected server cipher suites which don't support "
+        "Diffie-Hellman key exchange (DHE or ECDHE) i.e. in SERVER_HELLO "
+        "response")
+
+    def on_tls_response(self, record):
+        try:
+            self.log(logging.INFO, "NoForwardSecrecy::on_tls_response...")
+            for i, message in enumerate(record.messages):
+                # Check for Server Hello message
+                if (isinstance(message, tls.types.HandshakeMessage) and
+                        message.type == HandshakeMessage.TYPE.SERVER_HELLO):
+                    server_hello = message.obj
+                    selected_cipher = str(server_hello.cipher)
+                    _connection = self.connection
+                    destination = _connection.hostname if \
+                        _connection.hostname else _connection.server_addr
+                    debug_message = ["Selected cipher \"", selected_cipher,
+                            "\" for connection to \"", destination, "\""]
+                    self.log(logging.INFO, "".join(debug_message))
+                    """ Check if Ephemeral Diffie-Hellman key exchange is
+                        used in selected cipher """
+                    fs_key_strings = ["DHE", "ECDHE"]
+                    if not [fs_string for fs_string in fs_key_strings
+                            if fs_string in selected_cipher]:
```

Comment:
This is crazy hard to read...

Reply:
@RandellP I made that piece of code more readable. Pl review & merge whenever you get a chance. The other dependent PR on https://github.com/puppetlabs/pe_acceptance_tests/pull/1925 needs to be merged along with else  pe_acceptance_tests will start breaking (with just this). I have updated the 1925 to refer to puppetlabs/pe fork/branch of nogotofail. 
thanks,

Reply:
Actually pl hold. I realized this will break < 2017.x.x versions of pe_acceptance_tests because of the shared nogotofail repo branch being used across all pe_acceptance_tests. 
It is time we figure out and implement a way to deal with this like creating version branches etc. 
/cc @joshcooper PE versions earlier than 2017.3.x do not negotiate DH so this test needs to be  limited to 2017.3.x and above versions. Other way/option could be to update 2016.4.x version of pe_acceptance_tests to explicitly specify all data handlers so they don't end up invoking the Forward Secrecy handler. But that can be cumbersome and confusing.